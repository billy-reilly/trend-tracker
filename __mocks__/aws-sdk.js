/**
 * DynamoDB
 */
export const mockQuery = jest.fn();
export const mockPutItem = jest.fn();
export const mockUpdateItem = jest.fn();
export const mockDeleteItem = jest.fn();
export const mockScan = jest.fn();

const DynamoDB = jest.fn().mockImplementation(() => ({
  query: mockQuery,
  putItem: mockPutItem,
  updateItem: mockUpdateItem,
  deleteItem: mockDeleteItem,
  scan: mockScan
}));

/**
 * Lambda
 */
export const mockInvoke = jest.fn();

const Lambda = jest.fn().mockImplementation(() => ({
  invoke: mockInvoke
}));

/**
 * SDK interface
 */
const fakeAWSSDKInterface = {
  DynamoDB,
  Lambda
};

export default fakeAWSSDKInterface;
