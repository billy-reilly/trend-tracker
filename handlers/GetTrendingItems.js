import AWSXRay from "aws-xray-sdk-core";
import AWSSDK from "aws-sdk";

import configs from "../trend-list-configs";

const AWS = AWSXRay.captureAWS(AWSSDK);
const dynamodb = new AWS.DynamoDB();

const getTrendingItems = trendListId =>
  new Promise((resolve, reject) => {
    const config = configs[trendListId] || configs.default;
    dynamodb.query(
      {
        TableName: "InteractionCounts",
        IndexName: "trendListId-interactionCount-index",
        KeyConditionExpression: "trendListId = :tl",
        ExpressionAttributeValues: {
          ":tl": {
            S: trendListId
          }
        },
        ProjectionExpression: "itemId, interactionCount",
        Limit: config.trendListLimit,
        ScanIndexForward: false,
        ConsistentRead: true
      },
      (err, data) => {
        if (err) {
          reject(err);
        }
        resolve(data);
      }
    );
  });

export const handler = (event, context, cb) => {
  const { trendListId } = event.queryStringParameters;

  return getTrendingItems(trendListId)
    .then(data => {
      try {
        const responseBody = data.Items.reduce(
          (acc, item) => ({
            ...acc,
            [item.itemId.S]: parseInt(item.interactionCount.N, 10)
          }),
          {}
        );

        cb(null, {
          statusCode: 200,
          body: JSON.stringify(responseBody)
        });
      } catch (e) {
        cb(null, {
          statusCode: 500,
          body: `Error formatting responseBody: ${e.message}`
        });
      }
    })
    .catch(err => {
      return cb(null, {
        statusCode: 500,
        body: `Error reading from dynamodb: ${err.message}`
      });
    });
};
