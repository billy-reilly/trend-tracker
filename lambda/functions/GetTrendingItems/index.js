const AWSXRay = require("aws-xray-sdk-core");
const AWS = AWSXRay.captureAWS(require("aws-sdk"));
const dynamodb = new AWS.DynamoDB();

const config = {
  TS: {
    trendListLimit: "10",
    aggregationWindow: 1000 * 60 // 1 min in ms
  }
};

const getTrendingItems = catalogId =>
  new Promise((resolve, reject) => {
    dynamodb.query(
      {
        TableName: "ItemCounts",
        IndexName: "catalogId-eventCount-index",
        KeyConditionExpression: "catalogId = :cat",
        ExpressionAttributeValues: {
          ":cat": {
            S: catalogId
          }
        },
        ProjectionExpression: "id, eventCount",
        Limit: config[catalogId].trendListLimit,
        ScanIndexForward: false
      },
      (err, data) => {
        if (err) {
          reject(err);
        }
        resolve(data);
      }
    );
  });

exports.handler = (event, context, cb) => {
  if (!(event.queryStringParameters && event.queryStringParameters.catalogId)) {
    // TODO: handle invalid catalogId e.g. XX

    return cb(null, {
      statusCode: 400,
      body: JSON.stringify({
        message: "catalogId is a required query parameter"
      })
    });
  }

  return getTrendingItems(event.queryStringParameters.catalogId)
    .then(data => {
      try {
        const responseBody = data.Items.reduce(
          (acc, item) => ({
            ...acc,
            [item.id.S]: item.eventCount.N
          }),
          {}
        );

        cb(null, {
          statusCode: 200,
          body: JSON.stringify(responseBody)
        });
      } catch (e) {
        cb(null, {
          statusCode: 500,
          body: `Error formatting responseBody: ${e.message}`
        });
      }
    })
    .catch(err => {
      return cb(null, {
        statusCode: 500,
        body: `Error reading from dynamodb: ${err.message}`
      });
    });
};
