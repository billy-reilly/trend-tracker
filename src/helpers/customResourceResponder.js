// lifted from AWS cfn-response module source:
// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-lambda-function-code-cfnresponsemodule.html

export const SUCCESS = "SUCCESS";
export const FAILED = "FAILED";

export const send = (event, context, responseStatus, responseData) =>
  new Promise((resolve, reject) => {
    const responseBody = JSON.stringify({
      Status: responseStatus,
      Reason:
        "See the details in CloudWatch Log Stream: " + context.logStreamName,
      PhysicalResourceId: context.logStreamName,
      StackId: event.StackId,
      RequestId: event.RequestId,
      LogicalResourceId: event.LogicalResourceId,
      NoEcho: false,
      Data: responseData
    });

    const https = require("https");
    const url = require("url");

    const parsedUrl = url.parse(event.ResponseURL);
    const options = {
      hostname: parsedUrl.hostname,
      port: 443,
      path: parsedUrl.path,
      method: "PUT",
      headers: {
        "content-type": "",
        "content-length": responseBody.length
      }
    };

    const request = https.request(options, () => resolve());

    request.on("error", reject);

    request.write(responseBody);
    request.end();
  });
