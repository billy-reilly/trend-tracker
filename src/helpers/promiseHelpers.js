export const createPromiseCB = (resolve, reject) => (err, data) => {
  if (err) {
    return reject(err);
  }
  resolve(data);
};
