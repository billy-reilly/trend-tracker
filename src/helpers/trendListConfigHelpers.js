import AWSXRay from "aws-xray-sdk-core";
import AWSSDK from "aws-sdk";

import { createPromiseCB } from "../helpers/promiseHelpers";

const AWS = AWSXRay.captureAWS(AWSSDK);
const dynamodb = new AWS.DynamoDB();

const getTrendListConfigById = id =>
  new Promise((resolve, reject) => {
    dynamodb.getItem(
      {
        TableName: `${process.env.STACK_NAME}-TrendListConfigs`,
        Key: {
          trendListId: {
            S: id
          }
        }
      },
      createPromiseCB(resolve, reject)
    );
  }).then(data => {
    if (!data.Item) {
      throw new Error(`TrendListConfig not found for ${id}`);
    }

    let formattedData;
    try {
      formattedData = {
        trendListLimit: data.Item.trendListLimit.N,
        aggregationWindow: parseInt(data.Item.aggregationWindow.N, 10)
      };
    } catch (e) {
      throw new Error(
        `TrendListConfig data of unexpected shape for ${id}: ${e.message}`
      );
    }
    return formattedData;
  });

export const getTrendListConfig = id =>
  getTrendListConfigById(id).catch(err => {
    if (/^TrendListConfig not found for/.test(err.message)) {
      return getTrendListConfigById("default");
    }
    throw err;
  });
