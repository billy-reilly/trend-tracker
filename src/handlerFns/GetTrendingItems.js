import AWSXRay from "aws-xray-sdk-core";
import AWSSDK from "aws-sdk";

import { getTrendListConfig } from "../helpers/trendListConfigHelpers";
import { createPromiseCB } from "../helpers/promiseHelpers";

const AWS = AWSXRay.captureAWS(AWSSDK);
const dynamodb = new AWS.DynamoDB();

const getTrendingItems = (trendListId, trendListLimit) =>
  new Promise((resolve, reject) => {
    dynamodb.query(
      {
        TableName: `${process.env.STACK_NAME}-InteractionCounts`,
        IndexName: "trendListId-interactionCount-index",
        KeyConditionExpression: "trendListId = :tl",
        ExpressionAttributeValues: {
          ":tl": {
            S: trendListId
          }
        },
        ProjectionExpression: "itemId, interactionCount",
        Limit: trendListLimit,
        ScanIndexForward: false,
        ConsistentRead: true
      },
      createPromiseCB(resolve, reject)
    );
  });

export const handler = (event, context, cb) => {
  const { trendListId } = event.queryStringParameters;

  const promisedConfig = event.config
    ? Promise.resolve(event.config)
    : getTrendListConfig(trendListId);

  return (
    promisedConfig
      .then(({ trendListLimit }) =>
        getTrendingItems(trendListId, trendListLimit)
          .then(data => {
            try {
              const responseBody = data.Items.reduce(
                (acc, item) => ({
                  ...acc,
                  [item.itemId.S]: parseInt(item.interactionCount.N, 10)
                }),
                {}
              );

              cb(null, {
                statusCode: 200,
                body: JSON.stringify(responseBody)
              });
            } catch (e) {
              cb(null, {
                statusCode: 500,
                body: `Error formatting responseBody: ${e.message}`
              });
            }
          })
          .catch(err => {
            return cb(null, {
              statusCode: 500,
              body: `Error reading from dynamodb: ${err.message}`
            });
          })
      )
      // config error:
      .catch(err => {
        return cb(null, {
          statusCode: 500,
          body: err.message
        });
      })
  );
};
