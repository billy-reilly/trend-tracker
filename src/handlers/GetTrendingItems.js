import AWSXRay from "aws-xray-sdk-core";
import AWSSDK from "aws-sdk";

import { getTrendListConfig } from "../helpers/trendListConfigHelpers";
import { createPromiseCB } from "../helpers/promiseHelpers";

const AWS = AWSXRay.captureAWS(AWSSDK);
const dynamodb = new AWS.DynamoDB();

const getTrendingItems = (trendListId, trendListLimit) =>
  new Promise((resolve, reject) => {
    dynamodb.query(
      {
        TableName: "InteractionCounts",
        IndexName: "trendListId-interactionCount-index",
        KeyConditionExpression: "trendListId = :tl",
        ExpressionAttributeValues: {
          ":tl": {
            S: trendListId
          }
        },
        ProjectionExpression: "itemId, interactionCount",
        Limit: trendListLimit,
        ScanIndexForward: false,
        ConsistentRead: true
      },
      createPromiseCB(resolve, reject)
    );
  });

export const handler = (event, context, cb) => {
  // TODO when called from RecordInteraction pass down config and save extra DB reads

  const { trendListId } = event.queryStringParameters;

  // const configPromise = event.config
  //   ? getTrendListConfig(trendListId)
  //   : Promise.resolve(event.config);

  return (
    getTrendListConfig(trendListId)
      .then(({ trendListLimit }) =>
        getTrendingItems(trendListId, trendListLimit)
          .then(data => {
            try {
              const responseBody = data.Items.reduce(
                (acc, item) => ({
                  ...acc,
                  [item.itemId.S]: parseInt(item.interactionCount.N, 10)
                }),
                {}
              );

              cb(null, {
                statusCode: 200,
                body: JSON.stringify(responseBody)
              });
            } catch (e) {
              cb(null, {
                statusCode: 500,
                body: `Error formatting responseBody: ${e.message}`
              });
            }
          })
          .catch(err => {
            return cb(null, {
              statusCode: 500,
              body: `Error reading from dynamodb: ${err.message}`
            });
          })
      )
      // config error:
      .catch(err => {
        return cb(null, {
          statusCode: 500,
          body: err.message
        });
      })
  );
};
