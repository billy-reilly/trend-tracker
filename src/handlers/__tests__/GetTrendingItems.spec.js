import { handler as getTrendingItemsHandler } from "../GetTrendingItems";
import { mockQuery } from "../../../__mocks__/aws-sdk";
import { getTrendListConfig } from "../../helpers/trendListConfigHelpers";

jest.mock("aws-xray-sdk-core", () => ({
  captureAWS: a => a
}));
jest.mock("aws-sdk");
jest.mock("../../helpers/trendListConfigHelpers");

const noop = () => {};

const fakeContext = {};
const fakeEvent = {
  queryStringParameters: {
    trendListId: "shoes"
  }
};
const fakeQueryData = {
  Items: [
    {
      itemId: { S: "flip flops" },
      interactionCount: { N: "12345" }
    },
    {
      itemId: { S: "old boot" },
      interactionCount: { N: "4" }
    }
  ]
};

describe("GetTrendingItems", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe("when the TrendListConfig is successfully retrieved", () => {
    const fakeConfig = {
      trendListLimit: "10"
    };
    beforeEach(() => {
      getTrendListConfig.mockResolvedValue(fakeConfig);
    });

    it("should query DynamoDB", () => {
      let queryParams;
      mockQuery.mockImplementation((params, cb) => {
        queryParams = params;
        cb(null, fakeQueryData);
      });

      return getTrendingItemsHandler(fakeEvent, fakeContext, noop).then(() => {
        expect(mockQuery).toHaveBeenCalled();
        expect(queryParams).toMatchSnapshot();
      });
    });

    it("should correctly format the query data and respond", () => {
      mockQuery.mockImplementation((params, cb) => {
        cb(null, fakeQueryData);
      });
      const cbMock = jest.fn();
      return getTrendingItemsHandler(fakeEvent, fakeContext, cbMock).then(
        () => {
          expect(cbMock).toHaveBeenCalledWith(null, {
            statusCode: 200,
            body: JSON.stringify({
              "flip flops": 12345,
              "old boot": 4
            })
          });
        }
      );
    });

    it("should handle the case where the query data is of an unexpected shape", () => {
      const unexpectedQueryData = {
        Items: [
          {
            partitionKey: { S: "x" },
            attribute: { N: "1" }
          },
          {
            partitionKey: { S: "y" },
            attribute: { N: "2" }
          }
        ]
      };
      mockQuery.mockImplementation((params, cb) => {
        cb(null, unexpectedQueryData);
      });
      const cbMock = jest.fn();
      return getTrendingItemsHandler(fakeEvent, fakeContext, cbMock).then(
        () => {
          expect(cbMock).toHaveBeenCalledWith(null, {
            statusCode: 500,
            body:
              "Error formatting responseBody: Cannot read property 'S' of undefined"
          });
        }
      );
    });

    it("should handle the case where the db query fails", () => {
      const message = "ProvisionedThroughputExceededException or something";
      const fakeQueryError = new Error(message);
      mockQuery.mockImplementation((params, cb) => {
        cb(fakeQueryError);
      });
      const cbMock = jest.fn();
      return getTrendingItemsHandler(fakeEvent, fakeContext, cbMock).then(
        () => {
          expect(cbMock).toHaveBeenCalledWith(null, {
            statusCode: 500,
            body: `Error reading from dynamodb: ${message}`
          });
        }
      );
    });
  });

  describe("when there is an error retrieving the TrendListConfig", () => {
    const message = "Misconfigured...";
    const fakeConfigError = new Error(message);
    beforeEach(() => {
      getTrendListConfig.mockRejectedValue(fakeConfigError);
    });

    it("should respond with the error and not attempt to get the list of trending items", () => {
      const mockCB = jest.fn();
      return getTrendingItemsHandler(fakeEvent, fakeContext, mockCB).then(
        () => {
          expect(mockCB).toHaveBeenCalledWith(null, {
            statusCode: 500,
            body: message
          });
          expect(mockQuery).not.toHaveBeenCalled();
        }
      );
    });
  });
});
