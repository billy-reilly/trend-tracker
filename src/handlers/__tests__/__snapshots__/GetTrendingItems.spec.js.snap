// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetTrendingItems when the TrendListConfig is successfully retrieved should query DynamoDB 1`] = `
Object {
  "ConsistentRead": true,
  "ExpressionAttributeValues": Object {
    ":tl": Object {
      "S": "shoes",
    },
  },
  "IndexName": "trendListId-interactionCount-index",
  "KeyConditionExpression": "trendListId = :tl",
  "Limit": "10",
  "ProjectionExpression": "itemId, interactionCount",
  "ScanIndexForward": false,
  "TableName": "InteractionCounts",
}
`;
